firebase-firestore/src/main/java/com/google/firebase/firestore/remote/AbstractStream.java

```diff
@@ -109,11 +109,13 @@ public void onHeaders(Metadata headers) {
     public void onNext(RespT response) {
       dispatcher.run(
           () -> {
-            Logger.debug(
-                AbstractStream.this.getClass().getSimpleName(),
-                "(%x) Stream received: %s",
-                System.identityHashCode(AbstractStream.this),
-                response);
+            if (Logger.isDebugEnabled()) {
```

Comment:
Even with Proguard, logging adds about a second to 2-3 seconds the network usually takes.

Reply:
Hrm!  I wonder why...  Since we're not building up the log string or anything, it seems like this should just be a plain cheap method call (which immediately no-ops on the receiving end), unless `getClass().getSimpleName()` or `System.identityHashCode()` is expensive (my bet would be on the former).

Any interest in digging a little deeper to see why it's slow?  We may learn something that helps us preemptively improve other parts of the code [e.g. maybe we should cache this.getClass().getSimpleName()].

Reply:
Thanks for pushing me to dig into this. Looks like I got confused here. In the app, logging is on, which is why `String.format` is called here. When I tested the numbers, looks like I conflated no Proguard with logging, and Proguard with no logging (I used several repos to run tests, one of which had logging turned off).

To hopefully untangle this, I reran the numbers, using three "variables":
* SDK version;
* Proguard/no Proguard in the app;
* logging/no logging.

Results (three runs in each case, Release mode, same device as before):

* SDK 17.1.1, no Proguard, no logging: 4.4s - 4.8s (raw numbers: 4384ms / 4466ms / 4759ms);
* SDK 17.1.1, no Proguard, with logging: 4.5s - 4.8s (raw numbers: 4789ms / 4525ms / 4836ms);
* SDK 17.1.1, with Proguard, no logging: 4.4s - 4.6s (raw numbers: 4565ms / 4362ms / 4836ms);
* SDK 17.1.1, with Proguard, with logging: 4.5s - 4.6s (raw numbers: 4478ms / 4602ms / 4602ms (not a typo)).

So it seems that in 17.1.1 (IIUC, the last SDK version that was proguarded), the difference between logging enabled and disabled is negligible, and whether the app itself is proguarded doesn't really matter.

* SDK 17.1.2, no Proguard, no logging: 4.7s - 5.1s (raw numbers: 5064ms / 4992ms / 4717ms);
* SDK 17.1.2, no Proguard, with logging: 6.3s - 6.5s (raw numbers: 6303ms / 6351ms / 6503ms);
* SDK 17.1.2, with Proguard, no logging: 4.1s - 4.5s (raw numbers: 4361ms / 4094ms / 4490ms);
* SDK 17.1.2, with Proguard, with logging: 4.5s - 4.7s (raw numbers: 4536ms / 4503ms / 4682ms).

In 17.1.2, an app pays significant penalty if it enables logging but isn't proguarded. The rest of the numbers are probably within error margin.

* this branch, no Proguard, no logging: 2.7s - 2.9s (raw numbers: 2774ms / 2851ms / 2690ms);
* this branch, no Proguard, with logging: 4.8s - 4.9s (raw numbers: 4992ms / 4807ms / 4939ms);
* this branch, with Proguard, no logging: 2.7s - 2.8s (raw numbers: 2802ms / 2783ms / 2710ms);
* this branch, with Proguard, with logging: 2.8s - 2.9s (raw numbers: 2881ms / 2861ms / 2795ms).

Surprisingly, for this branch, the penalty of no Proguard/logging seems even higher than 17.1.2 (but perhaps within fluctuation). If Proguard is enabled, logging doesn't add any significant difference.